pkg/ccl/backupccl/system_schema.go

```diff
@@ -242,12 +245,105 @@ func usersRestoreFunc(
 	return nil
 }
 
+func roleMembersRestoreFunc(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	txn *kv.Txn,
+	systemTableName, tempTableName string,
+) (retErr error) {
+	if !execCfg.Settings.Version.IsActive(ctx, clusterversion.V23_1RoleMembersTableHasIDColumns) {
+		return defaultSystemTableRestoreFunc(ctx, execCfg, txn, systemTableName, tempTableName)
+	}
+
+	// It's enough to just check if role_id exists since member_id was added at
+	// the same time.
+	hasIDColumnsQuery := fmt.Sprintf(`SELECT EXISTS (SELECT 1 FROM [SHOW COLUMNS FROM %s] WHERE column_name = 'role_id')`, tempTableName)
+	row, err := execCfg.InternalExecutor.QueryRow(ctx, "has-id-columns", txn, hasIDColumnsQuery)
+	if err != nil {
+		return err
+	}
+	hasIDColumns := tree.MustBeDBool(row[0])
+	if hasIDColumns {
+		return defaultSystemTableRestoreFunc(ctx, execCfg, txn, systemTableName, tempTableName)
+	}
+
+	deleteQuery := fmt.Sprintf("DELETE FROM system.%s WHERE true", systemTableName)
+	log.Eventf(ctx, "clearing data from system table %s with query %q", systemTableName, deleteQuery)
+
+	_, err = execCfg.InternalExecutor.Exec(ctx, systemTableName+"-data-deletion", txn, deleteQuery)
+	if err != nil {
+		return errors.Wrapf(err, "deleting data from system.%s", systemTableName)
+	}
+
+	it, err := execCfg.InternalExecutor.QueryIteratorEx(ctx, systemTableName+"-query-all-rows",
+		txn, sessiondata.NodeUserSessionDataOverride,
+		fmt.Sprintf(`SELECT * FROM %s`, tempTableName),
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		retErr = errors.CombineErrors(retErr, it.Close())
+	}()
+
+	type insertStmtWithArgs struct {
+		stmt string
+		args []interface{}
+	}
+
+	var insertStmts []insertStmtWithArgs
+	for {
+		ok, err := it.Next(ctx)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			break
+		}
+
+		row := it.Cur()
+
+		userNames := map[string]tree.DString{
+			"role":   tree.MustBeDString(row[0]),
+			"member": tree.MustBeDString(row[1]),
+		}
+		isAdmin := tree.MustBeDBool(row[2])
+
+		ids := make(map[string]*tree.DOid)
+		for _, col := range []string{"role", "member"} {
+			row, err := execCfg.InternalExecutor.QueryRow(ctx, "get-user-id",
+				txn, `SELECT user_id FROM system.users WHERE username = $1`, userNames[col],
+			)
+			if err != nil {
+				return err
+			}
+			ids[col] = tree.MustBeDOid(row[0])
+		}
+
+		restoreQuery := fmt.Sprintf(`INSERT INTO system.%s ("role", "member", "isAdmin", role_id, member_id) VALUES ($1, $2, $3, $4, $5)`, systemTableName)
```

Comment:
❓ Any particular reason to keep N copies of this around when it is the same statement for every row?

Also same question as in the other code about whether you think it would be cleaner to get the role_id and member_id in this SQL statement.

Reply:
Good catch, I've refactored this to get the IDs with SQL like you suggested